brightener.cpp

```diff
@@ -3,40 +3,51 @@
 ImageBrightener::ImageBrightener(std::shared_ptr<Image> inputImage): m_inputImage(inputImage) {
 }
 
-int ImageBrightener::BrightenWholeImage() {
+bool ImageBrightener::BrightenWholeImage(int& attenuatedPixelCount) {
+    if (!m_inputImage->imageSizeIsValid()) {
+        return false;
+    }
+
     // For brightening, we add a certain grayscale (25) to every pixel.
     // While brightening, some pixels may cross the max brightness. They are
     // called 'attenuated' pixels
-    int attenuatedPixelCount = 0;
-    for (int x = 0; x < m_inputImage->m_rows; x++) {
-        for (int y = 0; y < m_inputImage->m_columns; y++) {
-            if (m_inputImage->pixels[x * m_inputImage->m_columns + y] > (255 - 25)) {
-                ++attenuatedPixelCount;
-                m_inputImage->pixels[x * m_inputImage->m_columns + y] = 255;
-            } else {
-                int pixelIndex = x * m_inputImage->m_columns + y;
-                m_inputImage->pixels[pixelIndex] += 25;
-            }
+    m_inputImage->pixelRunner([&attenuatedPixelCount](uint8_t pixelGrayscale, int pixelIndex) {
+        uint8_t outputGrayscale = pixelGrayscale;
+        if (outputGrayscale > (255 - 25)) {
+            ++attenuatedPixelCount;
+            outputGrayscale = 255;
         }
-    }
-    return attenuatedPixelCount;
+        else {
+            outputGrayscale += 25;
+        }
+
+        return outputGrayscale;
+        });
+    return true;
 }
 
-bool ImageBrightener::AddBrighteningImage(std::shared_ptr<Image> imageToAdd, int& attenuatedCount) {
+bool ImageBrightener::AddBrighteningImage(const std::shared_ptr<Image> imageToAdd, int& attenuatedPixelCount) {
+    if (!m_inputImage->imageSizeIsValid()) {
+        return false;
+    }
+
     if (imageToAdd->m_rows != m_inputImage->m_rows || imageToAdd->m_columns != m_inputImage->m_columns) {
         return false;
     }
-    attenuatedCount = 0;
-    for (int x = 0; x < m_inputImage->m_rows; x++) {
-        for (int y = 0; y < m_inputImage->m_columns; y++) {
-            int pixelIndex = x * m_inputImage->m_columns + y;
-            if (int(m_inputImage->pixels[pixelIndex]) + imageToAdd->pixels[pixelIndex] > 255) {
-                ++attenuatedCount;
-                m_inputImage->pixels[pixelIndex] = 255;
-            } else {
-                imageToAdd->pixels[pixelIndex] += m_inputImage->pixels[pixelIndex];
-            }
+
+    m_inputImage->pixelRunner([&attenuatedPixelCount, imageToAdd](uint8_t pixelGrayscale, int pixelIndex) {
```

Comment:
Good approach to capture `imageToAdd`. Makes it easy to get polymorphic behavior out of `pixelRunner` by altering its callbacks 👍